File: flutter/multando_sdk/lib/src/services/vehicle_type_service.dart
import '../core/http_client.dart';
import '../models/vehicle_type.dart';

/// Service for retrieving vehicle type reference data.
/// Results are cached in memory after the first fetch.
class VehicleTypeService {
  VehicleTypeService({required MultandoHttpClient httpClient})
      : _http = httpClient;

  final MultandoHttpClient _http;
  List<VehicleTypeResponse>? _cache;

  /// List all available vehicle types.
  Future<List<VehicleTypeResponse>> list({bool forceRefresh = false}) async {
    if (_cache != null && !forceRefresh) return _cache!;

    final response = await _http.get<dynamic>('/vehicle-types');
    final data = response.data;

    List items;
    if (data is Map<String, dynamic>) {
      items = data['items'] as List? ?? [];
    } else if (data is List) {
      items = data;
    } else {
      items = [];
    }

    _cache = items
        .cast<Map<String, dynamic>>()
        .map(VehicleTypeResponse.fromJson)
        .toList();
    return _cache!;
  }

  /// Invalidate the in-memory cache.
  void clearCache() {
    _cache = null;
  }
}

File: react-native/src/services/vehicleTypeService.ts
import { AxiosInstance } from 'axios';
import { Logger } from '../core/logger';
import { VehicleTypeResponse } from '../models/vehicleType';

export class VehicleTypeService {
  private http: AxiosInstance;
  private logger: Logger;
  private cache: VehicleTypeResponse[] | null = null;
  private cacheTimestamp: number = 0;
  private cacheTtl: number;

  constructor(http: AxiosInstance, logger: Logger, cacheTtlMs: number = 300000) {
    this.http = http;
    this.logger = logger;
    this.cacheTtl = cacheTtlMs;
  }

  async list(forceRefresh = false): Promise<VehicleTypeResponse[]> {
    if (!forceRefresh && this.cache && this.isCacheValid()) {
      this.logger.debug('Returning cached vehicle types');
      return this.cache;
    }

    this.logger.debug('Fetching vehicle types from API');
    const response = await this.http.get<VehicleTypeResponse[]>('/vehicle-types');
    this.cache = response.data;
    this.cacheTimestamp = Date.now();
    return this.cache;
  }

  getById(id: string): VehicleTypeResponse | undefined {
    return this.cache?.find((vt) => vt.id === id);
  }

  clearCache(): void {
    this.cache = null;
    this.cacheTimestamp = 0;
  }

  private isCacheValid(): boolean {
    return Date.now() - this.cacheTimestamp < this.cacheTtl;
  }
}
